KAFKA-19479: at_least_once mode in Kafka Streams silently drops messages when the producer fails with MESSAGE_TOO_LARGE, violating delivery guarantees

shashankhs11 opened the pull request:
Fixes a bug in producer.

Cf - https://github.com/apache/kafka/pull/20254#issuecomment-3363498212
and [KAFKA-19479](https://issues.apache.org/jira/browse/KAFKA-19479) for
more details

shashankhs11 committed:
add helper functions and result for tests

shashankhs11 committed:
test: added 3 passing tests

junrao commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java:
```diff
@@ -1062,6 +1062,146 @@ public void testSplitAndReenqueue() throws ExecutionException, InterruptedExcept
         assertEquals(1, future2.get().offset());
     }
 
+    // This test confirms us that splitting a single large record
```
We have the following code in Sender. It only tries to split the batch if it has more than 1 record. So, the splitting shouldn't occur forever.

```
    private void completeBatch(ProducerBatch batch, ProduceResponse.PartitionResponse response, long correlationId,
                               long now, Map<TopicPartition, Metadata.LeaderIdAndEpoch> partitionsWithUpdatedLeaderInfo) {
        Errors error = response.error;

        if (error == Errors.MESSAGE_TOO_LARGE && batch.recordCount > 1 && !batch.isDone() &&
                (batch.magic() >= RecordBatch.MAGIC_VALUE_V2 || batch.isCompressed())) {
            ...
            this.accumulator.splitAndReenqueue(batch);
```

junrao replied:
This test is not useful since we won't call split if a batch contains only 1 record. Ditto for testRetrySplitAndReenqueueBehaviourWithSingleLargeRecord().

shashankhs11 replied:
Cleaned up in 4762a99

junrao reviewed (commented):
@shashankhs11 : Thanks for the PR. Left a comment.

shashankhs11 committed:
patch to check record level futures

junrao commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -237,6 +238,20 @@ public boolean completeExceptionally(
+    /**
+     * Get all record futures for this batch.
+     * This is used by flush() to wait on individual records rather than the batch-level future.
+     * When batches are split, individual futures are chained to the new batches,
```
individual futures => individual record futures

shashankhs11 replied:
Done in 616e5fe

junrao commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -237,6 +238,20 @@ public boolean completeExceptionally(
+     * Get all record futures for this batch.
+     * This is used by flush() to wait on individual records rather than the batch-level future.
+     * When batches are split, individual futures are chained to the new batches,
+     * ensuring flush() waits for all split batches to complete.
```
ensuring flush() waits => ensuring that flush() waits

shashankhs11 replied:
Done in 616e5fe

junrao commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -1072,12 +1073,25 @@ private boolean appendsInProgress() {
+            // We first collect all futures into a list first to avoid holding references to
+            // ProducerBatch objects, allowing them to be garbage collected after completion.
+            List<FutureRecordMetadata> futures = new ArrayList<>();
+            for (ProducerBatch batch : this.incomplete.copyAll()) {
```
incomplete.requestResults() is no longer used?

shashankhs11 replied:
`requestResults()` doesn't expose record level futures. They are inside the private `thunks` list. 
That's why I used `copyAll()` to get the actual `ProducerBatch` objects directly and then extract the record futures. 
Is my approach correct?

junrao commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -1072,12 +1073,25 @@ private boolean appendsInProgress() {
+            // by waiting on record futures, we ensure flush() blocks until all split
+            // batches complete.
+            //
+            // We first collect all futures into a list first to avoid holding references to
```
first used twice in the same sentence.

shashankhs11 replied:
Done in 616e5fe

junrao commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java:
```diff
@@ -1066,6 +1066,146 @@ public void testSplitAndReenqueue() throws ExecutionException, InterruptedExcept
+
+    // here I am testing the hasRoomFor() behaviour
+    // It allows the first record no matter the size
+    // but does not allow the second record
```
This test is not useful. By design, we always allow the first record on a new batch.

shashankhs11 replied:
Cleaned up in 4762a99

junrao commented on streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/AtLeastOnceDeliveryMessageLossIntegrationTest.java:
```diff
@@ -0,0 +1,273 @@
+        streamsConfiguration.put(ProducerConfig.BATCH_SIZE_CONFIG, 33554432);
+    }
+
+    private void produceInputData(final int recordCount) throws Exception {
```
This method doesn't throw Exception.

shashankhs11 replied:
Removed in 38f9b9f

junrao commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:
```diff
@@ -237,6 +238,20 @@ public boolean completeExceptionally(
+     */
+    public List<FutureRecordMetadata> recordFutures() {
+        return thunks.stream()
+            .map(thunk -> thunk.future)
```
The downside of this approach is the potential overhead due to record level checking. Could you do some perf test to compare the flush() time with and w/o this PR with say a few thousands pending records?

Another option is to also chain ProduceRequestResult when splitting the batches. This allows us to check the status at the request level in the flush() call.

shashankhs11 replied:
> Another option is to also chain ProduceRequestResult when splitting the batches.

I checked this. We would then need to add chaining logic to `ProduceRequestResult` right?

shashankhs11 replied:
> Could you do some perf test to compare the flush() time with and w/o this PR with say a few thousands pending records?

Added a test in ae54bc2d692f7663242498ef94f014c1837cd798

I ran a performance test with 5000 pending records:
- Without PR (on avg): `0.163 ms`
- With PR (on avg): `1.436 ms`
- Increase: `+1.273 ms`

It seems to be significantly slower :smiling_face_with_tear:

junrao replied:
Ok, then it's probably useful to do the optimized approach. We will need to add a list of depending ProduceRequestResult in ProduceRequestResult. ProduceRequestResult.await() will further wait for the depending ProduceRequestResult after local latch is unblocked. In ProducerBatch.split(), we collect all ProduceRequestResult in batches and chain them to produceFuture.

shashankhs11 replied:
reverted previous fix in 07d545c and added optimized approach in 0bd4c99

junrao reviewed (commented):
@shashankhs11 : Thanks for the updated PR. A few more comments.

shashankhs11 committed:
remove exception

shashankhs11 committed:
fix typos

shashankhs11 committed:
remove useless tests

shashankhs11 committed:
add perf test

junrao commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java:
```diff
@@ -1790,4 +1825,56 @@ public void testSplitAndReenqueuePreventInfiniteRecursion() throws InterruptedEx
     }
+
+    @Test
+    public void testFlushPerformanceWithManyRecords() throws Exception {
```
Could we turn this into a jmh test and add it to the jmh-benchmarks directory?

shashankhs11 replied:
Yes! Done in aa1314f

junrao commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java:
```diff
@@ -1066,6 +1066,41 @@ public void testSplitAndReenqueue() throws ExecutionException, InterruptedExcept
         assertEquals(1, future2.get().offset());
     }
 
+    // here I am testing the hasRoomFor() behaviour
```
Could we add a unit test that fails without the fix?

shashankhs11 replied:
@junrao I would need your guidance in writing this unit test. This is what I tried:
- create an oversized batch to be split
- try to split the batch by calling `beginFlush()`, drain and then `splitAndReenqueue()`
- verify `originalBatch.produceFuture.completed()` returns true? -- Is this correct?
- then verify `recordFutures.isDone()` returns false
- then we can call `awaitFlushCompletion()` and check if it returns immediately?
- lastly verify that the record futures are still incomplete when the flush returns

Is my understanding and approach correct? Please let me know

junrao replied:
With the optimized approach, we could just 
1. create an oversized batch to be split
2. try to split the batch by calling splitAndReenqueue()
3. verify originalBatch.produceFuture.completed() returns false

shashankhs11 replied:
added in 5616a8f

junrao reviewed (commented):
@shashankhs11 : Thanks for the updated PR. A few more comments.

shashankhs11 committed:
jmh test

shashankhs11 committed:
add unit test

shashankhs11 committed:
revert previous fix

shashankhs11 committed:
add optimized fix with refactoring

junrao commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProduceRequestResult.java:
```diff
@@ -129,6 +148,17 @@ public TopicPartition topicPartition() {
+            return false;
+        }
+        // are all the dependent results completed?
+        synchronized (dependentResults) {
```
Hmm, this is not quite right. This method is called from FutureRecordMetadata. If a batch is split into two batches b1 and b2. For records in b1, they don't need to wait for the completion for b2. So, it doesn't seem that we should change this method. But it will be useful to document the semantic that it doesn't wait for dependent requests.
 
```
    @Override
    public boolean isDone() {
        if (nextRecordMetadata != null)
            return nextRecordMetadata.isDone();
        return this.result.completed();
    }

```

shashankhs11 replied:
@junrao I have a doubt regarding our test expectations:

> verify originalBatch.produceFuture.completed() returns false

This would then return `true`, right? Because `completed()` now would only check the latch which is released when the split occurs. So, we would then need to verify:

- `originalBatch.produceFuture.completed()` is `true`
- `future1.isDone()` is `false`
- `future2.isDone()` is `false`

So, we're testing that individual futures wait for their split batches (via chaining), not that the original batch's `completed()` waits for dependents. 

Is my understanding correct?

junrao replied:
@shashankhs11 : This is a bit confusing. I am suggesting multiple changes.
1. ProduceRequestResult.completed() currently is only called by FutureRecordMetadata.isDone(). Since FutureRecordMetadata doesn't need to wait for all splitted ProduceRequestResults to complete, we want ProduceRequestResult.completed() to stay as it is and only depend on the result before the split.
2. Introduce a new method like ProduceRequestResult.awaitAllDependants() and change flush() to call that.
3. Write a unit test that makes sure ProduceRequestResult.awaitAllDependants() waits until all dependents complete.

shashankhs11 replied:
Understood! Thank you so much for the detailed explanation, jun. I've made the suggested changes in e0d67a3

junrao commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProduceRequestResult.java:
```diff
@@ -70,7 +73,23 @@ public void done() {
+
+    /**
+     * Await the completion of this request.
+     * Note: This only waits for the local latch and not dependent results.
```
flush() actually calls await() and it needs to wait for all dependent requests to complete. Perhaps we could add a separate method like awaitAllDependants() and use that in flush().
```
    public void awaitFlushCompletion() throws InterruptedException {
        try {
            // Obtain a copy of all of the incomplete ProduceRequestResult(s) at the time of the flush.
            // We must be careful not to hold a reference to the ProduceBatch(s) so that garbage
            // collection can occur on the contents.
            // The sender will remove ProducerBatch(s) from the original incomplete collection.
            for (ProduceRequestResult result : this.incomplete.requestResults())
                result.await();
        } finally {
            this.flushesInProgress.decrementAndGet();
        }
    }
```

shashankhs11 replied:
added in e0d67a3

junrao commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProduceRequestResult.java:
```diff
@@ -33,6 +35,7 @@ public class ProduceRequestResult {
 
     private final CountDownLatch latch = new CountDownLatch(1);
     private final TopicPartition topicPartition;
+    private final List<ProduceRequestResult> dependentResults = new ArrayList<>();
```
This is an existing issue. There is a typo "record set was sent was sent".

shashankhs11 replied:
removed in 97e6f5e

junrao replied:
Could we add a comment to explain what dependentResults is?

shashankhs11 replied:
added in 4403920

junrao reviewed (commented):
@shashankhs11 : Thanks for the updated PR. A few more comments.

shashankhs11 committed:
fix typo

shashankhs11 committed:
new changes

junrao commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProduceRequestResult.java:
```diff
@@ -70,7 +75,29 @@ public void done() {
+     *
+     * @param dependentResult The dependent result to wait for
+     */
+    public void addDependentResult(ProduceRequestResult dependentResult) {
```
addDependentResult => addDependent ?

shashankhs11 replied:
done in 47bd6f5

junrao commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProduceRequestResult.java:
```diff
@@ -70,7 +75,29 @@ public void done() {
-     * Await the completion of this request
+     * Add a dependent ProduceRequestResult that must complete before this result is considered complete.
+     * This is used when a batch is split into multiple batches - the original batch's result
+     * should not complete until all split batches have completed.
```
>  the original batch's result should not complete until all split batches have completed.

In some cases like flush(), the original batch's result should not complete until all split batches have completed.

shashankhs11 replied:
added in 4403920

junrao commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java:
```diff
@@ -1790,4 +1828,65 @@ public void testSplitAndReenqueuePreventInfiniteRecursion() throws InterruptedEx
     }
+
+    @Test
+    public void testProduceRequestResultawaitAllDependents() throws Exception {
```
await => Await

shashankhs11 replied:
fixed in f5bb5fa

junrao commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java:
```diff
@@ -1790,4 +1828,65 @@ public void testSplitAndReenqueuePreventInfiniteRecursion() throws InterruptedEx
+            }
+        });
+        awaitThread.start();
+        Thread.sleep(100);
```
100 ms is too long for a unit test. How about 5ms?

shashankhs11 replied:
Done, changed to 5ms in f5bb5fa

junrao commented on streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/AtLeastOnceDeliveryMessageLossIntegrationTest.java:
```diff
@@ -0,0 +1,273 @@
+    public void shouldNotCommitOffsetsAndNotProduceOutputRecordsWhenProducerFailsWithMessageTooLarge() throws Exception {
+        
+        try (final LogCaptureAppender appender = LogCaptureAppender.createAndRegister(Sender.class)) {
+            produceInputData(LARGE_RECORD_COUNT);
```
produceInputData() doesn't produce large records.

shashankhs11 replied:
Hi @junrao, really sorry for the delayed response, and thank you for your patience. I just wanted to confirm my understanding:
- Currently, the test tests with a large number of records (50k) that creates a large batch
- but each record itself is not large.

> produceInputData() doesn't produce large records.

I am a bit unsure about the intended testing scenario. Should we also be testing with each of those 50k records being large (maybe 500KB - 1MB each)?

Please guide me on the scenario/s you would like me to test.

junrao replied:
@shashankhs11 : Thanks for the explanation. I understood the intention now. The test configured a large BATCH_SIZE_CONFIG in Streams producer to force a split. The test now fails for me locally.

```
org.opentest4j.AssertionFailedError: Consumer offsets should not be committed ==> 
Expected :true
Actual   :false
<Click to see difference>


	at org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
	at org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
	at org.junit.jupiter.api.AssertTrue.failNotTrue(AssertTrue.java:63)
	at org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:36)
	at org.junit.jupiter.api.Assertions.assertTrue(Assertions.java:214)
	at org.apache.kafka.streams.integration.AtLeastOnceDeliveryMessageLossIntegrationTest.shouldNotCommitOffsetsAndNotProduceOutputRecordsWhenProducerFailsWithMessageTooLarge(AtLeastOnceDeliveryMessageLossIntegrationTest.java:146)
```
I think this is because we fixed https://github.com/apache/kafka/pull/20358/files which removed the infinite loop during splitting. To force the producer error in the test, we will need to set [max.message.bytes](https://kafka.apache.org/documentation/#topicconfigs_max.message.bytes) to smaller than the produced record.

shashankhs11 replied:
> The test now fails for me locally.

The test was intended to fail initially to show that the bug exists. But at the time of writing the test, there was infinite loop behaviour which was later fixed by this - https://github.com/apache/kafka/pull/20358/files as you rightly said.
But the test shows that the offsets were still committed, which is still a bug, right?


> To force the producer error in the test, we will need to set [max.message.bytes](https://kafka.apache.org/documentation/#topicconfigs_max.message.bytes) to smaller than the produced record.

Should I add a new test to force the producer error? Just to confirm -- This would mimic the behaviour as the unit test that we wrote earlier, right?

junrao replied:
@shashankhs11 : In this test, since the producer can split the batch and republish the records successfully, so the stream job succeeds in the end. If I change 0 to LARGE_RECORD_COUNT in the following, the test passes, which is expected.

```
            final int outputRecordCount = verifyOutputRecords(LARGE_RECORD_COUNT); // should not produce records
            final boolean offsetsCommitted = verifyConsumerOffsetsCommitted(LARGE_RECORD_COUNT); // should not commit offset unless records are produced

            assertEquals(LARGE_RECORD_COUNT, outputRecordCount, "Output topic should not have any records");
```

Since the unit test covers the bug already, I am not sure that we need AtLeastOnceDeliveryMessageLossIntegrationTest.

shashankhs11 replied:
@junrao: Just a quick question for my understanding only

> If I change 0 to LARGE_RECORD_COUNT in the following, the test passes, which is expected.

 my understanding was that if offsets are committed, then the corresponding records should have been delivered to the output topic as well, right? Or just needs to be processed and not necessarily delivered?

> Since the unit test covers the bug already, I am not sure that we need AtLeastOnceDeliveryMessageLossIntegrationTest.

Removed it in 0e9c522

junrao replied:
Yes, if the offsets are committed, the output topic should have records. In the test above, the following test did succeed.

`assertEquals(LARGE_RECORD_COUNT, outputRecordCount, "Output topic should not have any records");`

shashankhs11 replied:
Understood! Thanks a lot for confirming.

junrao commented on streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/AtLeastOnceDeliveryMessageLossIntegrationTest.java:
```diff
@@ -0,0 +1,273 @@
+            produceInputData(LARGE_RECORD_COUNT);
+
+            kafkaStreams = createStreamsApplication();
+            startApplicationAndWaitUntilRunning(Collections.singletonList(kafkaStreams), Duration.ofMillis(DEFAULT_TIMEOUT));
```
How long does this test run? We want to use a reasonable timeout to avoid long running tests.

shashankhs11 replied:
With the current setup, it runs for about 30-31 seconds

junrao commented on streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/AtLeastOnceDeliveryMessageLossIntegrationTest.java:
```diff
@@ -0,0 +1,273 @@
+        try (final LogCaptureAppender appender = LogCaptureAppender.createAndRegister(Sender.class)) {
+            produceInputData(LARGE_RECORD_COUNT);
+
+            kafkaStreams = createStreamsApplication();
```
Where does the stream app create large records?

junrao reviewed (commented):
@shashankhs11 : Thanks for the updated PR. A few more comments.

shashankhs11 committed:
rename to addDependent

shashankhs11 committed:
fix typo in testname and change sleep time

shashankhs11 committed:
fixed and added more comment

junrao commented:
@shashankhs11 : Could you fix the build error? Also, could you run the same jmh test again to make sure there is no regression on the flush() time?

shashankhs11 committed:
fix build error

shashankhs11 commented:
# JMH Test Report

This is how I ran the test
1. I created a new local branch based off trunk with just the jmh test and ran the test.
2. Ran the test in this branch
3. 2 different cases - `numRecords = 5000` and `numRecords = 10000`

## Result BEFORE fix

<img width="1148" height="98" alt="image" src="https://github.com/user-attachments/assets/57937d6d-3a4c-4db4-95b2-69644bd41544" />

### Case 1: `numRecords = 5000`

<img width="1078" height="1053" alt="image" src="https://github.com/user-attachments/assets/150b6058-90d1-4e8b-8eeb-86b6685869b6" />

### Case 2: `numRecords = 10000`

<img width="1116" height="1095" alt="image" src="https://github.com/user-attachments/assets/36c1c866-ee3e-447d-9d8e-0497c2402e04" />

---

## Result AFTER fix (current branch)

<img width="1148" height="98" alt="image" src="https://github.com/user-attachments/assets/e28083c6-23ca-4d82-bce5-42d93739164d" />

### Case 1: `numRecords = 5000`

<img width="1078" height="1050" alt="image" src="https://github.com/user-attachments/assets/83cd60db-e429-464a-9bc1-812e7b55a607" />

### Case 2: `numRecords = 10000`

<img width="1089" height="886" alt="image" src="https://github.com/user-attachments/assets/a7f4694c-19df-4a12-bd29-956f9d448f4e" />

shashankhs11 commented:
@junrao: fixed the build error and added jmh test report in a seperate comment above. There seems to be a regression when `numRecords = 10000` -- Would this be acceptable?

Also left a comment in reply to this - https://github.com/apache/kafka/pull/20285#discussion_r2433587667

shashankhs11 committed:
fix build error in jmh

junrao commented:
@shashankhs11 : The small regression from 0.004 to 0.005 ms/op is negligible and is fine.

shashankhs11 committed:
remove test

shashankhs11 commented:
CI check with Java 25 seemed to be failing. Rebased with latest trunk.

junrao commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/ProduceRequestResult.java:
```diff
@@ -70,7 +82,29 @@ public void done() {
 
     /**
-     * Await the completion of this request
+     * Add a dependent ProduceRequestResult that must complete before this result is considered complete.
```
Add a dependent ProduceRequestResult that must complete before this result is considered complete. => Add a dependent ProduceRequestResult.

shashankhs11 replied:
Done in [896ecc1](https://github.com/apache/kafka/pull/20285/commits/896ecc1a28a6cf3df1e7a78f82791a03d3ef7cea)

junrao reviewed (commented):
@shashankhs11 : Thanks for the updated PR. Just a minor comment. Also, could you rebase to pick up a flaky test fix https://github.com/apache/kafka/pull/20713 ?

shashankhs11 committed:
fix minor comment

shashankhs11 commented:
> Also, could you rebase to pick up a flaky test fix

@junrao: Done :)

junrao reviewed (approved):
@shashankhs11 : Thanks for the updated PR. LGTM. Could you update the description of the PR before merging?

shashankhs11 commented:
> Could you update the description of the PR before merging?

@junrao: Done! Changed the title as well as the description.

Thank you very much for your time and patience, Jun. I am very very grateful for the opportunity to have worked closely with you over the past couple of weeks. I got to learn a lot from this :)

junrao commented:
@shashankhs11 : I kept the PR name to match the jira name, but adjusted the description in the commit message. Thanks for working on this issue!

mjsax commented:
Thanks for reviewing this PR @junrao!

mjsax commented:
Cherry-picked to `4.1`, `4.0`, and `3.9` branches.
